File: blind75/55_longest_palindromic_substring/longest_palindromic_substring.go
package longest_palindromic_substring

// LongestPalindrome returns the longest palindromic substring in s.
// Uses expand around center approach for O(n²) time complexity.
func LongestPalindrome(s string) string {
	if len(s) < 2 {
		return s
	}

	start, maxLen := 0, 0

	for i := 0; i < len(s); i++ {
		// Check for odd-length palindromes (center at i)
		len1 := expandAroundCenter(s, i, i)

		// Check for even-length palindromes (center between i and i+1)
		len2 := expandAroundCenter(s, i, i+1)

		// Get the maximum length from both cases
		length := max(len1, len2)

		// Update if we found a longer palindrome
		if length > maxLen {
			maxLen = length
			start = i - (length-1)/2
		}
	}

	return s[start : start+maxLen]
}

// expandAroundCenter expands around the center and returns the length of the palindrome.
func expandAroundCenter(s string, left, right int) int {
	for left >= 0 && right < len(s) && s[left] == s[right] {
		left--
		right++
	}
	return right - left - 1
}

// max returns the maximum of two integers.
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: blind75/55_longest_palindromic_substring/longest_palindromic_substring.py
def longest_palindrome(s: str) -> str:
    """
    Returns the longest palindromic substring in s.
    Uses expand around center approach for O(n²) time complexity.

    Args:
        s: Input string

    Returns:
        Longest palindromic substring
    """
    if len(s) < 2:
        return s

    start, max_len = 0, 0

    for i in range(len(s)):
        # Check for odd-length palindromes (center at i)
        len1 = expand_around_center(s, i, i)

        # Check for even-length palindromes (center between i and i+1)
        len2 = expand_around_center(s, i, i + 1)

        # Get the maximum length from both cases
        length = max(len1, len2)

        # Update if we found a longer palindrome
        if length > max_len:
            max_len = length
            start = i - (length - 1) // 2

    return s[start:start + max_len]


def expand_around_center(s: str, left: int, right: int) -> int:
    """
    Expands around the center and returns the length of the palindrome.

    Args:
        s: Input string
        left: Left pointer
        right: Right pointer

    Returns:
        Length of the palindrome found
    """
    while left >= 0 and right < len(s) and s[left] == s[right]:
        left -= 1
        right += 1

    return right - left - 1
